jni/lib/src/jni_object.dart

```diff
@@ -0,0 +1,143 @@
+import 'dart:ffi';
+
+import 'package:ffi/ffi.dart';
+
+import 'jni_bindings_generated.dart';
+import 'extensions.dart';
+import 'jni_class.dart';
+import 'jvalues.dart';
+
+part 'jni_object_methods_generated.dart';
+
+/// JniObject is a convenience wrapper around a JNI local object reference.
+///
+/// It holds the object, its associated associated jniEnv etc..
+/// It should be distroyed with [delete] method after done.
+///
+/// It's valid only in the thread it was created.
+/// When passing to code that might run in a different thread (eg: a callback),
+/// consider obtaining a global reference and reconstructing the object.
+class JniObject {
+  JClass _cls;
+  final JObject _obj;
+  final Pointer<JniEnv> _env;
+  JniObject.of(this._env, this._obj, this._cls);
+
+  JniObject.fromJObject(Pointer<JniEnv> env, JObject obj)
+      : _env = env,
+        _obj = obj,
+        _cls = nullptr;
+
+  /// Reconstructs a JniObject from [r]
+  ///
+  /// [r] still needs to be explicitly deleted when
+  /// it's no longer needed to construct any JniObjects.
+  JniObject.fromGlobalRef(Pointer<JniEnv> env, JniGlobalObjectRef r)
+      : _env = env,
+        _obj = env.NewLocalRef(r._obj),
+        _cls = env.NewLocalRef(r._cls);
+
+  /// Delete the local reference contained by this object.
+  ///
+  /// Do not use a JniObject after calling [delete].
```

Comment:
What happens if I make two `JniObject`s out of one `JObject`, and I call `delete` on one of them? Is that safe? In other words, does the `delete` only delete the `JniObject` object? Or does it also delete the underlying things?

Reply:
It deletes the underlying reference, i.e original JObject.

I think a warning in documentation is enough. Since there aren't many times you would want to convert a JObject (that's just raw `Pointer<Void>`) -> JniObject. And even then Jni.wrap() is enough.

Safer semantics can be implemented by creating new LocalRef whenever we are copying like this. But that will also prevent us from doing things like `Jni.wrap(low_level_expression_returning_jobject)`, because we will lose original LocalRef and that will leak.

Reply:
Let's wait and see what we generate for end users of JNIgen. If we can hide these implementation details from them, we only have to make sure that we generate the right thing ourselves. If we cannot fully hide this API we should come up with good documentation and a good API to manage these native resources.

Please file a bug for us to look into this later.